### ISSUE TITLE
Changing stride throws reshape error in contextual attention function

### ISSUE THREAD

**Issue by theahura** [JiahuiYu/generative_inpainting#185]
Hi Jiahui, 

Was looking at the implementation of the contextual attention layer, and I think I spotted a bug.

Stride is used here: 
`w = tf.extract_image_patches(b, [1,ksize,ksize,1], [1,stride,stride,1], [1,1,1,1], padding='SAME')`
and here:
`m = tf.extract_image_patches(mask, [1,ksize,ksize,1], [1,stride,stride,1], [1,1,1,1], padding='SAME')`
These lines extract a certain number of channels that is inversely proportional to the stride. 

Attention is extracted here: 
`yi = tf.nn.conv2d(xi, wi_normed, strides=[1,1,1,1], padding="SAME")`
and reshaped here:
`yi = tf.reshape(yi, [1, fs[1]*fs[2], bs[1]*bs[2], 1])`

The shapes fs and bs are not changed by stride at all. However, the number of channels in w and m, and therefore in yi, decreases inversely to the stride. As a result, if the stride is anything other than 1, the reshape in the fuse section will fail because there are not enough elements in the tensor.

A quick test on my end with a stride set to 3 failed. Let me know what the correct solution is for this!

**Comment by JiahuiYu** [comment]
@theahura I don't think it is a bug. If I remember correctly (currently I am completely on another project), the shapes fs and bs are not changed by stride **by design**. Can you try to use even stride set, like 2, 4, etc.?

Another option to reduce memory cost is to directly reshape fs and bs before attention, and reshape back after contextual attention. However, I think the option of striding is generally better.

**Comment by theahura** [comment]
(For reference I was testing this using my own implementation of deepfill2; I imagine you are busy getting ready to submit that work, so feel free to respond here when you have more spare cycles!)

An even stride also failed. 

fs and bs represent the shapes of the foreground and background; however, they are the same value because the same input is passed in for both f and b. 
`x, offset_flow = contextual_attention(x, x, mask_s, 3, 1, rate=2)`

As a result, this line:
`yi = tf.reshape(yi, [1, fs[1]*fs[2], bs[1]*bs[2], 1])`
really reshapes yi into a square of the same side length. 

When stride is set to N > 1, the output of tf.extract_image_patches() no longer has the same height and width. For example, in this line: 
`w = tf.extract_image_patches(b, [1,ksize,ksize,1], [1,stride,stride,1], [1,1,1,1], padding='SAME')`
If stride is set to 2, the height and width of w is correspondingly cut in half. 

After some number of reshapes, the number of channels ends up being stride^2 less than what is needed for the attention. 

I believe that the foreground shape fs needs to be the output shape of tf.extract_image_patches().

**Comment by JiahuiYu** [comment]
@theahura Good findings. Do you have any suggestion on how to fix the problem? I agree current implementation of contextual attention is (1) difficult to read and (2) may contain bug if hyper-parameters are changed.
